python/adelphi/bin/adelphi

```diff
@@ -103,104 +151,119 @@ def unique(iterable, key=None):
                 yield element
 
 
-def with_cluster(cluster_fn, ctx):
-    return _with_cluster(cluster_fn, **({k:ctx.obj[k] for k in ["hosts","port","username","password"]}))
+def with_cluster(cluster_fn, props):
+    return _with_cluster(cluster_fn, **({k:props[k] for k in ["hosts", "port", "username", "password"]}))
 
 
-def get_keyspaces(ctx, cluster):
-    keyspace_names = ctx.obj["keyspace_names"]
+def build_keyspace_id(ks):
+    m = hashlib.sha256()
+    m.update(ks.name.encode("utf-8"))
+    # Leverage the urlsafe base64 encoding defined in RFC 4648, section 5 to provide an ID which can
+    # safely be used for filenames as well
+    return urlsafe_b64encode(m.digest()).decode('ascii')
 
+
+def get_keyspaces(props, cluster):
+    keyspace_names = props["keyspace-names"]
     metadata = cluster.metadata
     keyspaces = build_keyspace_objects(keyspace_names, metadata)
 
-    return [anonymize_keyspace(ks) for ks in keyspaces] if ctx.obj['anonymize'] else keyspaces
-
-
-def validate_keyspaces(keyspace_objs):
-    if len(keyspace_objs) == 0:
-        log.info("No keyspaces selected")
+    if len(keyspaces) == 0:
+        log.info("No valid keyspaces selected")
         exit(1)
 
+    log.info("Processing the following keyspaces: %s", ','.join((ks.name for ks in keyspaces)))
+    return {anonymize_keyspace(ks) if props['anonymize'] else ks : build_keyspace_id(ks) for ks in keyspaces}
+
 
 def get_cluster_metadata(cluster):
     hosts = cluster.metadata.all_hosts()
     unique_dcs = unique((host.datacenter for host in hosts))
-    return {"host_count": len(hosts), "dc_count": sum(1 for i in unique_dcs)}
-
-
-def gather_keyspaces(ctx):
-    partial_fn = partial(get_keyspaces, ctx)
-    keyspaces = with_cluster(partial_fn, ctx)
-
-    log.info("Exporting schema for keyspaces %s", ','.join((k.name for k in keyspaces)))
+    return {"host-count": len(hosts), "dc-count": sum(1 for i in unique_dcs)}
 
-    validate_keyspaces(keyspaces)
-    return keyspaces
 
+def build_keyspace_map_and_metadata(props):
+    def build_fn(cluster):
+        metadata = {k : props[k] for k in ["purpose", "maturity"] if k in props}
+        metadata.update(get_cluster_metadata(cluster))
+        metadata["creation_timestamp"] = datetime.now(utc).isoformat()
```

Comment:
A timestamp for schema creation has general utility.  This was being created in the Github contribute logic but I promoted it here so that every schema will include it as metadata.